File: app/src/main/java/de/mchllngr/quickopen/service/StartApplicationService.java
package de.mchllngr.quickero.service;

import android.app.IntentService;
import android.content.Intent;
import android.os.Bundle;

import com.f2prateek.rx.preferences.Preference;
import com.f2prateek.rx.preferences.RxSharedPreferences;

import androidx.preference.PreferenceManager;
import de.mchllngr.quickero.R;
import de.mchllngr.quickero.util.CustomNotificationHelper;
import de.mchllngr.quickero.util.FirebaseUtils;

/**
 * {@link IntentService} for starting the clicked application from the notification.
 */
public class StartApplicationService extends IntentService {

    public StartApplicationService() {
        super(StartApplicationService.class.getName());
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        // get packageName and start selected application
        Bundle extras = intent.getExtras();
        if (extras != null && extras.containsKey(getString(R.string.key_package_name))) {
            String packageName = extras.getString(getString(R.string.key_package_name));
            startApplication(packageName);
        }

        // close navigation bar
        Intent i = new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS);
        sendBroadcast(i);

        // check if version is still supported
        FirebaseUtils.isVersionSupported(supported -> {
            handleVersionSupportedResult(supported);

            // stop service
            stopSelf();
        });
    }

    /**
     * Starts the application with the given package-name.
     *
     * @param packageName package-name from selected application
     */
    private void startApplication(String packageName) {
        FirebaseUtils.logStartApplicationEvent(this);
        Intent launchIntent = getPackageManager().getLaunchIntentForPackage(packageName);
        if (launchIntent != null)
            startActivity(launchIntent);
    }

    private void handleVersionSupportedResult(boolean supported) {
        if (!supported) {
            // set notification disabled
            RxSharedPreferences rxSharedPreferences = RxSharedPreferences.create(PreferenceManager.getDefaultSharedPreferences(this));
            Preference<Boolean> notificationEnabledPref = rxSharedPreferences.getBoolean(
                    getString(R.string.pref_notification_enabled),
                    Boolean.parseBoolean(getString(R.string.pref_notification_enabled_default_value))
            );
            notificationEnabledPref.set(false);

            // show notification
            new CustomNotificationHelper(this).showVersionNotSupportedNotification();
        }
    }
}

File: app/src/main/java/de/mchllngr/quickero/service/StartApplicationService.kt
package de.mchllngr.quickero.service

import android.annotation.SuppressLint
import android.app.IntentService
import android.content.Intent
import android.os.Build

/** [IntentService] for starting the clicked application from the notification. */
class StartApplicationService : IntentService(StartApplicationService::class.java.name) {

    override fun onHandleIntent(intent: Intent?) {
        // get packageName and start selected application
        val extras = intent?.extras
        if (extras != null && extras.containsKey(KEY_PACKAGE_NAME)) {
            val packageName = extras.getString(KEY_PACKAGE_NAME)
            startApplication(packageName)
        }

        @SuppressLint("MissingPermission")
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            // close navigation bar
            val i = Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS)
            sendBroadcast(i)
        }
    }

    private fun startApplication(packageName: String?) {
        if (packageName.isNullOrBlank()) return
        val launchIntent = packageManager.getLaunchIntentForPackage(packageName)
        launchIntent?.let { startActivity(it) }
    }

    companion object {

        const val KEY_PACKAGE_NAME = "package_name"
    }
}
